### PR TITLE
118 ledger support

### PR DESCRIPTION
**Pull request by AbstractFruitFactory**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -66,6 +66,7 @@
     "node": ">=8.9"
   },
   "dependencies": {
+    "@ledgerhq/hw-transport-node-hid": "^5.12.0",
```
Comment (MuncleUscles) [inline package.json:69]: Is this going to work in the browser?

File: src/modules/identity/RadixLedgerIdentity.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+} from './ledgerAppTypes'
+
+const CLA = 0xaa
+
+// HD Derivation Path: /0'/0/0
+const BIP44_PATH = '80000000' + '00000000' + '00000000'
+
+const CHUNK_SIZE = 250
+
+export default class RadixLedgerIdentity implements RadixIdentity {
```
Comment (MuncleUscles) [inline src/modules/identity/RadixLedgerIdentity.ts]: I think the naming is confusing, seems like it might refer to the RadixLedger component

File: src/modules/identity/RadixLedgerIdentity.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    private device
+    public address: RadixAddress
+    public account: RadixAccount
+
+    public async init() {
```
Comment (MuncleUscles) [inline src/modules/identity/RadixLedgerIdentity.ts]: This is problematic, it means you could have an uninitialised identity would fail to sign stuff.

I think it would be better to handle any initialisation in a static factory method, like RemoteIdentity

File: src/modules/identity/RadixIdentity.ts [hunk 4]
```diff
@@ -23,18 +23,18 @@
...
-    public account: RadixAccount
+export default interface RadixIdentity extends RadixSignatureProvider, RadixDecryptionProvider {
+    signAtom(atom: RadixAtom): Promise<RadixAtom>
+    decryptECIESPayload(payload: Buffer): Promise<Buffer>
+    decryptECIESPayloadWithProtectors(protectors: Buffer[], payload: Buffer): Promise<Buffer>
+    getPublicKey(): Buffer
+    account: RadixAccount
 }
+
+export function setupIdentityAccount(this, address: RadixAddress): RadixAccount {
```
Comment (MuncleUscles) [inline src/modules/identity/RadixIdentity.ts]: Why this change from a constructor to a function, and from a class to an interface? It's odd, since you need this function in every type of identity, that's what classes are for

File: package.json (hunk 1 above)
Comment (AbstractFruitFactory) [reply package.json:69]: no, Martin said to only support the desktop wallet for now. Just have to make sure it doesn't break in the browser

File: src/modules/identity/RadixIdentity.ts [hunk 5]
```diff
@@ -23,18 +23,18 @@
...
-        this.account.enableDecryption(this)
-    }
-
-    public abstract signAtom(atom: RadixAtom): Promise<RadixAtom>
-    public abstract decryptECIESPayload(payload: Buffer): Promise<Buffer>
-    public abstract decryptECIESPayloadWithProtectors(protectors: Buffer[], payload: Buffer): Promise<Buffer>
-    public abstract getPublicKey(): Buffer
-
-    public account: RadixAccount
+export default interface RadixIdentity extends RadixSignatureProvider, RadixDecryptionProvider {
```
Comment (AbstractFruitFactory) [inline src/modules/identity/RadixIdentity.ts]: RadixLedgerIdentity could not directly inherit from RadixIdentity, since initiating the RadixAccount is an asynchronous action (need to get public key from ledger).
When hierarchies start to fall apart, better use composition I think. Hence the refactor...

File: package.json (hunk 1 above)
Comment (MuncleUscles) [reply package.json:69]: In electron, you have 2 environments - the main thread is node.js, and the windows are effectively Chrome. I'm not a 100% sure, but I think the stuff running in the BrowserWindows, which this will have to be in, is effectively a browser environment. I think it makes sense to check at least

Comment (AbstractFruitFactory) [reply package.json:69]: ahh interesting ... that's very good to know. Will definitely test in the wallet environment to make sure. I know at least that some ledger transports are compatible with browser and some are for node. Just need to test it out.

File: src/modules/identity/RadixIdentity.ts (hunk 4 above)
Comment (AbstractFruitFactory) [reply src/modules/identity/RadixIdentity.ts]: Ok you win... but I like functions :cry:

File: src/modules/identity/RadixLedgerIdentity.ts (hunk 2 above)
Comment (AbstractFruitFactory) [reply src/modules/identity/RadixLedgerIdentity.ts]: Yeah I was thinking about this. Might use HardwareWallet instead?

File: src/modules/identity/RadixLedgerIdentity.ts (hunk 3 above)
Comment (AbstractFruitFactory) [reply src/modules/identity/RadixLedgerIdentity.ts]: will fix

File: src/modules/identity/RadixIdentity.ts (hunk 4 above)
Comment (MuncleUscles) [reply src/modules/identity/RadixIdentity.ts]: Functions are great, but I think in this case it's kind of awkward to have to know that you need to import this alongside the interface

File: src/modules/identity/RadixLedgerIdentity.ts (hunk 2 above)
Comment (MuncleUscles) [reply src/modules/identity/RadixLedgerIdentity.ts]: We might implement other types of hardware wallets in the future, maybe HardwareWallerLedgerNano (if this would work for all types of ledger, otherwise LedgeNanoS even)

Comment (AbstractFruitFactory) [reply src/modules/identity/RadixLedgerIdentity.ts]: How about HWLedger? as in Hardware Wallet Ledger
